src/passes/OptimizeInstructions.cpp

```diff
@@ -570,6 +570,22 @@ struct OptimizeInstructions
         return ret;
       }
     } else if (auto* unary = curr->dynCast<Unary>()) {
+      if (unary->op == EqZInt32 || unary->op == EqZInt64) {
```

Comment:
Could you move this up to the top of this function with the other matcher-based transformations? The more we can avoid putting new logic in this highly-nested part of the function, the better. It would also be good to start the match with the top-level expression, even though we already know it is a unary expression. Moving this will also let you use the `binary` matcher without the `Match::` prefix.

Reply:
I decide use slightly different approach due to input wasm could contain mixed `eqz` and `x == 0` operations and due to `x == 0` --> `eqz x` transform apply in early stage exactly after canonicalization I guess better just remove `x == 0` matching and always check only `x != 0`. So `x == 0` --> `eqz x` may be make sense even move to `canonicalization` routine.

Reply:
I still think it would be good to move these up to where the other matcher-based transformations are. It's ok if you have to move other code as well to make it work out. I would also be willing to do a non-functional change to port most of this code to use matchers first if that would make it easier for you.

Reply:
Done!